### ISSUE TITLE
suggests/imports contradictions, and tcltk in particular

### ISSUE THREAD

**Issue by jarioksa** [vegandevs/vegan#84]
**Vegan** currently suggests two functions in its `DESCRIPTION`: **tcltk** and **parallel**. Simultaneously **vegan** imports their namespace in its `NAMESPACE` file. The idea of suggestion is that the package is not essential to **vegan**, but is only needed in special functions and loaded (by `require()`) only when needed. However, the package namespace is loaded with import. This means that loading of **vegan** will fail if the package is not available -- which is against the idea of only suggesting the package that is not necessarily needed.

We only suggest and import recommended and base packages and we have assumed this can be done safely. However, because all namespace imports are needed, we should not list those packages as suggests but as imports in `DESCRIPTION`. We had a specific discussion on **tcltk** when we split off **vegan3d**, but as **tcltk** is a base package, we assumed it is safe to have in **vegan**. However, **tcltk** is dependent on external system utilities (Tcl/Tk) that may not exist and having **tcltk** is a configurable option in **R**. If **R** has been compiled without **tcltk** capability, installing current **vegan** will fail:
```sh
** preparing package for lazy loading
Warning: S3 methods ‘as.character.tclObj’, ‘as.character.tclVar’, ‘as.double.tclObj’, 
‘as.integer.tclObj’, ‘as.logical.tclObj’, ‘as.raw.tclObj’, ‘print.tclObj’, ‘[[.tclArray’, ‘[[<-.tclArray’, 
‘$.tclArray’, ‘$<-.tclArray’, ‘names.tclArray’, ‘names<-.tclArray’, ‘length.tclArray’, ‘length<-.tclArray’, 
‘tclObj.tclVar’, ‘tclObj<-.tclVar’, ‘tclvalue.default’, ‘tclvalue.tclObj’, ‘tclvalue.tclVar’, ‘tclvalue<-.default’, 
‘tclvalue<-.tclVar’, ‘close.tkProgressBar’ were declared in NAMESPACE but not found
Error : .onLoad failed in loadNamespace() for 'tcltk', details:
  call: fun(libname, pkgname)
  error: Tcl/Tk support is not available on this system
ERROR: lazy loading failed for package ‘vegan’
* removing ‘/home/jarioksa/R/x86_64-unknown-linux-gnu-library/3.2/vegan’
* restoring previous ‘/home/jarioksa/R/x86_64-unknown-linux-gnu-library/3.2/vegan’
```
If you manage to get through the installation process, you will get the same error with `library(vegan)`. This should not happen with a suggested package.

I will suggest a PR #85  that will fix this issue. It will remove  `import(tcltk)`  from the `NAMESPACE` and prefix every Tcl/Tk command with `tcltk::`. I tried to avoid this when making the changes for new namespace rules because there are 126 `tcltk::` commands, and changing all these was boring and required several iterations to seek all needed cases.

**parallel** is another suggested package. It is a recommended package, and a lean **R** can be built without recommended packages. I don't expect such a lean **R** to contain **vegan** so that we may pretty safely assume that it has recommended packages if it has **vegan**. However, there is now a contradiction between `DESCRIPTION` (which suggests **parallel**) and `NAMESPACE` (which compulsorily imports **parallel**) and this could be changed. Either we
1. change `DESCRIPTION` to import **parallel** and remove `require(parallel)` in eleven functions (because it will be loaded anyway), or
2. only suggest **parallel**, but remove `import(parallel)` from `NAMESPACE`  and seek and prefix all **parallel** commands with `parallel::`  in functions.

Currently we import selectively from **cluster**, **mgcv** and **MASS** which means that these packages must be available to install and load **vegan**. All are recommended packages, but could be only suggested. I think `cluster::daisy()`  would be the easiest case in three files. We use **mgcv** only in `ordisurf`, but there prefixing as `mgcv::gam(y ~ mgcv::s())` looks ugly. **MASS** usage is more dispersed, but could also be replaced with `MASS::` prefixed commands.

**Comment by jarioksa** [comment]
`R CMD check` has changed after I supposed to have fixed **tcltk** calls. Now we get this message in R (2015-01-05 r67326):
```
* checking dependencies in R code ... NOTE
'library' or 'require' calls in package code:
  ‘parallel’ ‘tcltk’
  Please use :: or requireNamespace() instead: see 'Suggested Packages'.
```
This needs checking and will delay the release of 2.2-1 that I had planned today,

**Comment by gavinsimpson** [comment]
Won't prefixing with `::` *alone* just delay errors with **tcltk** functions? It would be better to do as the Writing R Extensions manual says and use:

```r
if (requireNamespace("rgl", quietly = TRUE)) {
   rgl::plot3d(...)
} else {
   ## do something else not involving rgl.
}
```

that way we can quickly fail with an informative error message if **tcltk** is not available.

As for Recommended packages; I think we can safely include these; yes you can build R without them but most people using **vegan** won't be building their own R. I see **tcltk** a little differently as it does require additional system libraries which aren't always available on Mac OS X; hence you can have R without having **tcltk** functionality and may not be able to install/set this up easily. Linus users have easy access to tcl and tk devel packages, and with Windows IIRC, this is part of the Windows R installer. Recommended packages are easy to install via dependencies with tools users will already be using (i.e. `install.packages()`)

**Comment by gavinsimpson** [comment]
I suggest we `importFrom()`

 * MASS
 * cluster
 * mgcv
 * parallel

and leave **tcltk** as a suggest-ed package but follow the advice from Writing R Extensions to `requireNamespace` and prefix with `::`, the former so we can catch problems and bail out early with an informative error message.

**Comment by jarioksa** [comment]
Yep, it must be `requireNamespace`. I'll fix that. However, I'm now going for `parallel::` idiom, and I'd like to change `cluster::daisy()` to that, too.

**Comment by gavinsimpson** [comment]
Just `parallel::` or in cbomination with `requireNamespace()`? Don't  we run into the same problem if `parallel` (or whatever pkg) is not available?

Why do you favour the prefix instead of selectively importing? These aren't obscure packages and I don't see the harm in being more explicit in requiring them (via `Imports`)?

**Comment by jarioksa** [comment]
`DESCRIPTION` and usage must be consistent. If we `Suggest` a package, we cannot `import` it because then package **must** be available (and that is pretty Corleone style of suggestion). I think this is clear with **tcltk** and I have now changed that function so that I have replaced `require(tcltk) || stop()` with `requireNamespace("tcltk") || stop()`.   I think it is clear that **tcltk** must be suggested and we should be able to use **vegan** without **tcltk**. However, **parallel** is a more intricate issue. I think I still have atavistic memories of early history when **parallel** was under development, and could trigger crashes with some work flows and **R** versions. That is the reason why we only `Suggest` **parallel**. I have now (mostly) changed the **parallel**ized functions after the `Suggests` idiom. Alternatively, I could change the `DESCRIPTION` to `Imports: parallel` (not selectively: we use so many **parallel** functions that we can import all). The only differences would be in the `NAMESPACE` and dropping `requireNamespace("parallel")` from functions. Currently that `requireNamespace`  only check that **parallel** is available and will trigger serial processing if unavailable. With the `Imports` change in `DESCRIPTION` and `NAMESPACE` that test would be unnecessary because **vegan** would fail without **parallel**. 

The double-colons `::` should be useful even if we `Import` the functions, because they take care that we really use functions of that package instead of some other function with the same name earlier in the search path. Actually, we should have some more of those `::` to avoid obscure errors like this:

```R
> library(vegan)
Loading required package: permute
Loading required package: lattice
This is vegan 2.3-0
> data(mite, mite.env)
> m <- cca(mite)
> library(gam)
Loading required package: splines
Loaded gam 1.09.1

> ordisurf(m ~ WatrCont, mite.env)
Error in qr.lm(object) : lm object does not have a proper 'qr' component.
 Rank zero or should not have used lm(.., qr=FALSE).
```
